File: javascript/leetcode/linked-list-cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
    if (head === null) { return false; }
    var slow = head;
    var fast = head.next;
    while (fast !== slow) {
      if (slow === null || fast === null) {
        return false;
      }
      slow = slow.next;
      fast = fast.next !== null ? fast.next.next : fast.next;
    }
    return true;
};

File: java/leetcode/linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        HashMap<Integer, Integer> candidates = new HashMap<Integer, Integer>();
        while (head != null) {
            if (candidates.containsKey(head.val)) {
                return true;
            } else {
                candidates.put(head.val, 1);
            }
            head = head.next;
        }
        return false;
    }
}